File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/converter/DefaultTestStepResultConverter.java
package org.skellig.teststep.processing.converter;

import java.util.HashMap;
import java.util.Map;

public class DefaultTestStepResultConverter implements TestStepResultConverter {

    private Map<String, TestStepResultConverter> testStepResultConverters;

    protected DefaultTestStepResultConverter(Map<String, TestStepResultConverter> testStepResultConverters) {
        this.testStepResultConverters = testStepResultConverters;
    }

    @Override
    public Object convert(String convertFunction, Object result) {
        if (testStepResultConverters.containsKey(convertFunction)) {
            return testStepResultConverters.get(convertFunction).convert(convertFunction, result);
        } else {
            return result;
        }
    }

    @Override
    public String getConvertFunctionName() {
        return "";
    }


    public static class Builder {

        private Map<String, TestStepResultConverter> testStepResultConverters;

        public Builder() {
            testStepResultConverters = new HashMap<>();
        }

        public Builder withTestStepResultConverter(TestStepResultConverter testStepResultConverter) {
            this.testStepResultConverters.put(testStepResultConverter.getConvertFunctionName(), testStepResultConverter);
            return this;
        }

        public TestStepResultConverter build() {
            testStepResultConverters.put("string", new TestStepResultToStringConverter());

            return new DefaultTestStepResultConverter(testStepResultConverters);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/converter/DefaultTestStepResultConverter.kt
package org.skellig.teststep.processing.converter

import org.skellig.teststep.processing.value.function.FunctionValueExecutor


class DefaultTestStepResultConverter private constructor(val testStepResultConverters: Map<String, TestStepResultConverter>) :
    TestStepResultConverter,
    FunctionValueExecutor {

    override fun execute(name: String, args: Array<Any?>): Any? {
        return (testStepResultConverters[name] as FunctionValueExecutor?)?.execute(name, args)
    }

    override fun getFunctionName(): String = ""

    override fun convert(convertFunction: String, result: Any?): Any? {
        return if (testStepResultConverters.containsKey(convertFunction)) {
            testStepResultConverters[convertFunction]?.convert(convertFunction, result)
        } else {
            result
        }
    }

    override fun getConvertFunctionName(): String {
        return ""
    }

    class Builder {

        private val testStepResultConverters = mutableMapOf<String, TestStepResultConverter>()

        fun withTestStepResultConverter(testStepResultConverter: TestStepResultConverter) =
            apply { testStepResultConverters[testStepResultConverter.getConvertFunctionName()] = testStepResultConverter }

        fun build(): TestStepResultConverter {
            testStepResultConverters["string"] = TestStepResultToStringConverter()
            return DefaultTestStepResultConverter(testStepResultConverters)
        }
    }
}